File: code/SoundPlayground/SoundPlayground/Parser/AbstractSyntaxTree/MusicParallelNode.cs
using System.Collections.Generic;
using System.Linq;
using SoundPlayground.Core;
using SoundPlayground.VirtualMachine;

namespace SoundPlayground.Parser.AbstractSyntaxTree
{
    public class MusicParallelNode : MusicNode
    {
        public IEnumerable<MusicNode> Expressions { get; set; }

        public MusicParallelNode ( IEnumerable<MusicNode> nodes ) {
            Expressions = nodes;
        }

        public override IEnumerable<Note> GetCommands ( Context context ) {
            List<Context> forks = new List<Context>();

            IEnumerable<Note> notes = Expressions.Select( s => {
                Context forked = context.Fork();

                forks.Add( forked );

                return s.GetCommands( forked );
            } ).MergeSorted( note => note.Start );

            foreach ( Note note in notes ) {
                yield return note;
            }

            context.Join( forks.ToArray() );
        }
    }
}

File: code/SoundPlaygroundPy/parser/abstract_syntax_tree/music_parallel_node.py
from core import Music
from .music_node import MusicNode
from py_linq import Enumerable

class MusicParallelNode( MusicNode ):
    def __init__ ( self, nodes, position : (int, int) = None ):
        super().__init__( position )
        
        self.expressions = nodes
    
    def fork_and_get_events ( self, node, forks, context ):
        forked = context.fork()

        forks.append( forked )

        value = node.eval( forked )

        if isinstance( value, Music ):
            return value.expand( forked )

        return []

    def get_events ( self, context ):
        forks = []

        # TODO add merge_sorted extension method to enumerables
        notes = Enumerable( self.expressions )\
            .select( lambda node: self.fork_and_get_events( node, forks, context ) )\
            .merge_sorted( lambda note: note.timestamp )

        try:
            for note in notes:
                yield note
        finally:
            context.join( *forks )
        